metrics/metricstest/metricstest.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2019 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metricstest
+
+import (
+	"testing"
+
+	"go.opencensus.io/stats/view"
+)
+
+// CheckStatsReported checks that there is a view registered with the given name for each string in names,
+// and that each view has at least one record.
+func CheckStatsReported(t *testing.T, names ...string) {
+	t.Helper()
+	for _, name := range names {
+		d, err := view.RetrieveData(name)
+		if err != nil {
+			t.Errorf("Reporter.Report() error = %v, wantErr %v", err, false)
+		}
+		if len(d) < 1 {
+			t.Error("No data reported when data was expected, view data is empty.")
+		}
+	}
+}
+
+// CheckStatsNotReported checks that there are no records for any views that a name matching a string in names.
+// Names that do not match registered views are considered not reported.
+func CheckStatsNotReported(t *testing.T, names ...string) {
+	t.Helper()
+	for _, name := range names {
+		if d, err := view.RetrieveData(name); err != nil {
```

Comment:
Is this check backwards? We're checking that `err != nil`, but then proceeding to use `d` when `err` was **non-`nil`**.

Reply:
Its confusing to look at, but it should be ok.

We want this function to pass in two cases:
1. err == nil (implies no component registered a stat with "name", so it's not possible that it's reported)
2. if err != nil && len(d) == 0 (there is a stat registered with "name", but nothing actually recorded a measurement for that stat)

Reply:
I still think this is backwards. Let me rewrite, with comments:

```golang
d, err := view.RetrieveData(name)
// At this point, we have two cases:
// 1. d has data, and err is nil
// 2. d is nil, and err is non-nil
if err != nil { // We are in case 2
  if len(d) >0 {
    ....
  }
}